interfaces/builtin/gpio.go

```diff
@@ -0,0 +1,150 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"fmt"
+	"os"
+	"strconv"
+
+	"github.com/snapcore/snapd/interfaces"
+)
+
+var gpioSysfsGpioBase = "/sys/class/gpio/gpio"
+var gpioSysfsExport = "/sys/class/gpio/export"
+
+// GpioInterface type
+type GpioInterface struct{}
+
+// String returns the same value as Name().
+func (iface *GpioInterface) String() string {
+	return iface.Name()
+}
+
+// Name of the GpioInterface
+func (iface *GpioInterface) Name() string {
+	return "gpio"
+}
+
+// SanitizeSlot checks the slot definition is valid
+func (iface *GpioInterface) SanitizeSlot(slot *interfaces.Slot) error {
+	// Paranoid check this right interface type
+	if iface.Name() != slot.Interface {
+		panic(fmt.Sprintf("slot is not of interface %q", iface))
+	}
+
+	// We will only allow creation of this type of slot by a gadget or OS snap
+	if !(slot.Snap.Type == "gadget" || slot.Snap.Type == "os") {
+		return fmt.Errorf("gpio slots only allowed on gadget or core snaps")
+	}
+
+	// Must have a GPIO number
+	number, ok := slot.Attrs["number"]
+	if !ok {
+		return fmt.Errorf("gpio slot must have a number attribute")
+	}
+
+	// Valid values of number
+	if _, ok := number.(int); !ok {
+		return fmt.Errorf("gpio slot number attribute must be an int")
+	}
+
+	// Slot is good
+	return nil
+}
+
+// SanitizePlug checks the plug definition is valid
+func (iface *GpioInterface) SanitizePlug(plug *interfaces.Plug) error {
+	// Make sure right interface type
+	if iface.Name() != plug.Interface {
+		panic(fmt.Sprintf("plug is not of interface %q", iface))
+	}
+
+	// Plug is good
+	return nil
+}
+
+// PermanentPlugSnippet returns security snippets for plug at install
+func (iface *GpioInterface) PermanentPlugSnippet(plug *interfaces.Plug, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	switch securitySystem {
+	case interfaces.SecurityDBus, interfaces.SecurityAppArmor, interfaces.SecuritySecComp, interfaces.SecurityUDev, interfaces.SecurityMount:
+		return nil, nil
+	default:
+		return nil, interfaces.ErrUnknownSecurity
+	}
+}
+
+// ConnectedPlugSnippet returns security snippets for plug at connection
+func (iface *GpioInterface) ConnectedPlugSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	switch securitySystem {
+	case interfaces.SecurityDBus, interfaces.SecuritySecComp, interfaces.SecurityUDev, interfaces.SecurityMount:
+		return nil, nil
+	case interfaces.SecurityAppArmor:
+		path := fmt.Sprint(gpioSysfsGpioBase, slot.Attrs["number"])
+		// Entries in /sys/class/gpio for single GPIO's are just symlinks
+		// to their correct device part in the sysfs tree. As AppArmor
+		// does not handle symlinks we need to dereference the GPIO
```

Comment:
This should be rephrased-- "does not handle symlinks" sounds like a deficiency in AppArmor when in fact AppArmor does handle symlinks-- it resolves them because it needs to in order to prevent trivial breakouts.